File: android/lib/repository/src/main/kotlin/com/warrenbrowse/vpn/lib/repository/PurchaseClaim.kt
package com.warrenbrowse.vpn.lib.repository

import java.security.MessageDigest
import java.security.SecureRandom

/**
 * What the app keeps to collect a purchase it opened in the browser (warren-core doc 35
 * section 14). The [wpid] names the purchase and rides in the checkout URL, so it proves
 * nothing on its own. The pull secret is what warren-api asks for before it hands out the
 * voucher: only its SHA-256 goes to the checkout (the `ph` parameter), and the secret itself
 * leaves the app only inside the pull, through [code].
 */
class PurchaseClaim internal constructor(val wpid: String, private val pullSecret: String) {
    /** SHA-256 of the raw secret bytes, the digest warren-api stores. */
    val pullSecretHash: String
        get() = MessageDigest.getInstance("SHA-256").digest(pullSecret.hexToBytes()).toHex()

    /**
     * The wpid followed by the pull secret (96 hex chars): the form `WarrenJni.redeemVoucher`
     * recognizes as a purchase to collect rather than a voucher to redeem.
     */
    val code: String
        get() = wpid + pullSecret

    /** The checkout page for this purchase; the account chip fragment is the caller's. */
    fun checkoutUrl(base: String): String = "$base?pid=$wpid&ph=$pullSecretHash"

    // The secret must never reach a log line through string interpolation.
    override fun toString(): String = "PurchaseClaim([redacted])"

    companion object {
        /** A fresh 128-bit wpid and 256-bit pull secret. */
        fun mint(): PurchaseClaim {
            val random = SecureRandom()
            return PurchaseClaim(random.hex(16), random.hex(32))
        }

        private fun SecureRandom.hex(bytes: Int): String =
            ByteArray(bytes).also { nextBytes(it) }.toHex()
    }
}

private fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }

private fun String.hexToBytes(): ByteArray =
    ByteArray(length / 2) { i -> substring(2 * i, 2 * i + 2).toInt(16).toByte() }

File: desktop/packages/mullvad-vpn/src/main/purchase-claim.ts
import { createHash, randomBytes } from 'crypto';

// What the app keeps to collect a purchase it opened in the browser
// (warren-core doc 35 section 14). The wpid names the purchase and travels
// in the checkout URL, so it proves nothing on its own. The pull secret is
// what warren-api asks for before it hands out the voucher or the renewal
// handoff: only its SHA-256 goes to the checkout (the `ph` query
// parameter), and the secret itself travels only in the pull's body.
export interface PurchaseClaim {
  wpid: string;
  secret: string;
}

const CLAIM_CODE_RE = /^([0-9a-f]{32})([0-9a-f]{64})$/;

export function mintPurchaseClaim(): PurchaseClaim {
  return {
    wpid: randomBytes(16).toString('hex'),
    secret: randomBytes(32).toString('hex'),
  };
}

// SHA-256 of the raw secret bytes: the digest warren-api stores.
export function pullSecretHash(secretHex: string): string {
  return createHash('sha256').update(Buffer.from(secretHex, 'hex')).digest('hex');
}

// The form the daemon's submitVoucher takes for a purchase: wpid then
// secret, 96 hex characters. A voucher code (16 Crockford-32 characters)
// can never take that shape, so the daemon dispatches on it.
export function claimCode(claim: PurchaseClaim): string {
  return `${claim.wpid}${claim.secret}`;
}

export function parseClaimCode(code: string): PurchaseClaim | undefined {
  const match = CLAIM_CODE_RE.exec(code);
  return match ? { wpid: match[1], secret: match[2] } : undefined;
}
